docs/design/2021-09-22-data-consistency.md

```diff
@@ -0,0 +1,121 @@
+
+- Read errors: errors are reported when reading or writing inconsistent data
+- Data error: data can be read but the result is incorrect
+- Data Loss: Data written cannot be read
+
+It is very difficult for support engineers to locate data index inconsistencies when they occur.
+
+- Multiple possible causes of data errors
+- Inconsistency phenomenon cannot easily imply its cause
+- Difficulty in collecting information available for investigation on the cause
+
+In short, the serious impact of the problem and the difficulty of troubleshooting is the main reason why we need to invest in improvement at present.
+
+## Detailed Design
+
+There are two methods to reduce data corruption from happening and spreading.
+
+### Assertion
+
+Based on the operator characteristics and the information that the operator has already checked, some assertions can be made regarding to the data being mutated. The assertions are pushed down to TiKV and are executed before writing. DML operators (e.g. UpdateExec, InsertExec) are encapsulated in tables to make KV modifications to memDB:
+
+- Put KV: Add or update KV pairs
+- Del KV: delete KV pair
+
+While the data is modified, the KV pair modified to memDB is additionally set with 4 new Assertion Flags (occupying a total of 2 bits):
+
+- None(00): There is currently no Assertion for KV, but subsequent modifications in the transaction can overwrite this value [default Flag]
+- Unknown(11): KV pair cannot be Assertioned, and subsequent modifications within the transaction cannot be overwritten
+- MustExist(10): Key must exist in storage, subsequent modifications within the transaction cannot be overwritten
+- MustNotExist(01): Key must not exist in the storage, subsequent modifications within the transaction cannot be overwritten
+
+The flags in MemDB, like the KV data, are first set to the current stage of the statement. After that, if the statement reports an error and the statement rolls back, the flags will be invalidated. If the statement is executed successfully and the is submitted, the flags will take effect in the transaction. The assertion information is sent to TiKV in the prewrite stage of 2PC. TiKV performs additional checks based on the assertions before modifying data. If the assertions fail, TiKV rejects the request.
+
+### Check Mutations
+
+The basic idea is to ensure that the transaction will not write inconsistent data when the data before the transaction is consistent.
+
+For possible errors in the executor: Assume that the set of row values before and after the transaction are V1 and V2 respectively, and the data indices are consistent before the transaction is executed. The condition of data consistency after the transaction's commitment is V1-V2 = { Del_{index}.value }, V2-V1 = { Put_{index}.value }. 
```

Comment:
It should be possible for DBAs to check for consistency too. I know of a customer issue where both TiKV and TiFlash have data corruption, but it's not reported by `ADMIN CHECK TABLE`. So for TiFlash I created: https://github.com/pingcap/tidb/issues/27915

Reply:
Yeah enhancing `admin check` and `admin recover` should be part of the larger project in data inconsistencies. The proposal itself focuses on "reporting error in time".